File: 211DesignAddandSearchWordsDataStructure/DesignAddandSearchWordsDataStructure.py


import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
from node_structure.TrieNode import TrieNode

class DesignAddandSearchWordsDataStructure:
    def __init__(self):
        self.root = TrieNode()
    

    def addWord(self, word: "str") -> "None":
        cur = self.root
        for c in word:
            if not cur.children[ord(c) - ord('a')]:
                cur.children[ord(c) - ord('a')] = TrieNode()
            cur = cur.children[ord(c) - ord('a')]
        cur.end = True
    
    def search(self, word: "str") -> "bool":
        return self.__searchHelper(word, 0, self.root)

    def __searchHelper(self, word: "str", idx: "int", root: "TrieNode") -> "bool":
        cur = root
        for i in range(idx, len(word)):
            if word[i] == '.':
                for child in cur.children:
                    if child and self.__searchHelper(word, i + 1, child):
                        return True
                return False
            else:
                if cur.children[ord(word[i]) - ord('a')]:
                    cur = cur.children[ord(word[i]) - ord('a')]
                else:
                    return False
        return cur.end


def main():
    test = DesignAddandSearchWordsDataStructure()

if __name__ == "__main__":
    main()

File: 211DesignAddandSearchWordsDataStructure/DesignAddandSearchWordsDataStructure.cpp


#include <TrieNode.hpp>
#include <string>
#include <cstdlib>

class DesignAddandSearchWordsDataStructure {
private:
    TrieNode* root;


public:
    DesignAddandSearchWordsDataStructure(): root(new TrieNode()) {}

    ~DesignAddandSearchWordsDataStructure() {
        delete this->root; // hate rule of three :D
    }

    DesignAddandSearchWordsDataStructure(
        const DesignAddandSearchWordsDataStructure& rhs
    ) {
        // level order traversal?
    }


    void addWord(std::string word) {
        TrieNode* cur = this->root;
        for (size_t i = 0; i < word.size(); ++i) {
            char c = word[i];
            if (NULL == cur->children[c - 'a']) {
                cur->children[c - 'a'] = new TrieNode();
            }
            cur = cur->children[c - 'a'];
        }
        cur->end = true;
    }

    bool search(std::string word) {
        return searchHelper(word, 0, this->root);
    }
    
private:
    bool searchHelper(std::string& word, size_t idx, TrieNode* root) {
        TrieNode* cur = root;
        for (size_t i = idx; i < word.size(); ++i) {
            char c = word[i];
            if (c == '.') {
                for (size_t j = 0; j < 26; ++j) {
                    if (
                        NULL != cur->children[j] &&
                        searchHelper(word, i + 1, cur->children[j])
                    ) {
                        return true;
                    }
                }
                return false;
            } else {
                // actual word
                if (NULL == cur->children[c - 'a']) {
                    return false;
                } else {
                    cur = cur->children[c - 'a'];
                }
            }
        }
        return cur->end;
    }

    DesignAddandSearchWordsDataStructure& operator=(
        DesignAddandSearchWordsDataStructure& rhs
    ) {
        // eee
    }
};
